core/emitter.js

```diff
@@ -3,25 +3,43 @@ import logger from './logger';
 
 let debug = logger('quill:events');
 
-const EVENTS = ['selectionchange', 'mousedown', 'mouseup', 'click'];
-
-EVENTS.forEach(function(eventName) {
-  document.addEventListener(eventName, (...args) => {
-    [].slice.call(document.querySelectorAll('.ql-container')).forEach((node) => {
-      // TODO use WeakMap
-      if (node.__quill && node.__quill.emitter) {
-        node.__quill.emitter.handleDOM(...args);
-      }
-    });
-  });
-});
+const EVENTS = [ 'selectionchange', 'mousedown', 'mouseup', 'click' ];
+
+const setGlobalListeners = (emitter, documentOrShadow) => {
+  const selectionRoot = resolveSelectionRoot(documentOrShadow);
+
+  return EVENTS.reduce((acc, eventName) => {
+    const handler = (...args) => emitter.handleDOM(...args);
+
+    // set the listener on the correctly scoped document. This allows callers to pass the correct document to the emitter b/c document can be window.document or shadowRoot
+    // In addition, this fixes a hack where the quill instance was being stored in the DOM instead of closure scope.
+    selectionRoot.addEventListener(eventName, handler);
+
+    return acc.concat({ eventName, target: selectionRoot, handler });
+  }, []);
+};
+
+const resolveSelectionRoot = (selectionRoot) => {
+  if (selectionRoot && typeof selectionRoot.getSelection === 'function') {
+    return selectionRoot;
+  } else if (selectionRoot) {
+    return selectionRoot.ownerDocument || window.document;
+  }
 
+  return window.document;
+}
 
 class Emitter extends EventEmitter {
-  constructor() {
+  constructor({ selectionRoot = window.document }) {
     super();
     this.listeners = {};
     this.on('error', debug.error);
+    this.selectionRoot = selectionRoot;
+    this.globalListeners = setGlobalListeners(this, selectionRoot);
+  }
+
+  destroy() {
+    this.globalListeners.forEach(({ eventName, target, handler }) => target.removeEventListener(eventName, handler));
```

Comment:
For the case where someone uses this globally in plainJS, the destroy function needs to be called. 

* Update docs to require users to call destroy
* Setup some kind of detach listener which calls destroy

This is probably worth a conversation.